### PR TITLE
fix: ensure azure-arc-acstor namespace resource gets captured when fault tolerance is enabled

### PR DESCRIPTION
**Pull request by Elsie4ever**
1. ensure azure-arc-acstor namespace resource gets captured when fault tolerance is enabled
2. for CRD without namespace property, the resources will be captured under provided fallback namespace
image

---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to Azure IoT Operations tooling!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

Intent for Production

- [ ] It is expected that pull requests made to default or core branches such as `dev` or `main` are of production grade. Corollary to this, any merged contributions to these branches may be deployed in a public release at any given time. By checking this box, you agree and commit to the expected production quality of code.

Basic expectations

- [ ] If introducing new functionality or modified behavior, are they backed by unit and/or integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** the relevant unit **and** integration tests pass? i.e. `pytest <project root> -vv`. Please provide evidence in the form of a screenshot showing a succesful run of tests locally OR a link to a test pipeline that has been run against the change-set.
- [ ] Have linter checks passed using the `.pylintrc` and `.flake8` rules? Look at the CI scripts for example usage.
- [ ] Have extraneous print or debug statements, commented out code-blocks or code-statements (if any) been removed from the surface area of changes?
- [ ] Have you made an entry in HISTORY.rst which concisely explains your user-facing feature or change?

Azure IoT Operations CLI maintainers reserve the right to enforce any of the outlined expectations.

A PR is considered **ready for review** when all basic expectations have been met (or do not apply).

### CODE REVIEW

**Comment by Elsie4ever** [comment]
image

File: azext_edge/edge/providers/support/base.py [hunk 1]
```diff
@@ -69,12 +69,17 @@ def process_crd(
...
-        if not namespace:
+        # Try to get namespace from metadata, if not found, use fallback_namespace if provided
+        try:
             namespace = r["metadata"]["namespace"]
-        namespaces.append(namespace)
+        except KeyError:
+            if fallback_namespace:
+                namespace = fallback_namespace
+            else:
+                raise ValueError("Namespace not found in CRD metadata and no fallback namespace provided.")
```
Comment (digimaun) [inline azext_edge/edge/providers/support/base.py:80]: Support bundle should not crash, but with the way the support bundle runner works I believe this exception becomes a logger .debug message, please confirm

File: azext_edge/edge/providers/edge_api/arccontainerstorage.py [hunk 2]
```diff
@@ -10,3 +10,7 @@
 ARCCONTAINERSTORAGE_API_V1 = EdgeResourceApi(
     group="arccontainerstorage.azure.net", version="v1", moniker="arccontainerstorage"
```
Comment (digimaun) [inline azext_edge/edge/providers/edge_api/arccontainerstorage.py:11]: both these api groups exist at the same time?

Comment (digimaun) [reply azext_edge/edge/providers/edge_api/arccontainerstorage.py:11]: Interesting that docs show only usage of https://learn.microsoft.com/en-us/azure/azure-arc/container-storage/install-edge-volumes?tabs=multi

File: azext_edge/tests/edge/support/create_bundle_int/test_arccontainerstorage_int.py [hunk 3]
```diff
@@ -59,8 +62,55 @@ def test_create_bundle_arccontainerstorage(init_setup, tracked_files):
...
+        "acstor-cert-manager",
+        "acstor-crd",
+        "acstor-csi",
+        "acstor-etcd",
+        "acstor-io",
+        "acstor-ndm",
+        "acstor-operator",
+        "acstor-prereq",
+        "acstor-scripts",
+        "acstor-support-bundle",
```
Comment (vilit1) [inline azext_edge/tests/edge/support/create_bundle_int/test_arccontainerstorage_int.py:96]: at this point - can you just use acstor? or will something else that we don't want get captured?

File: azext_edge/edge/providers/edge_api/arccontainerstorage.py [hunk 4]
```diff
@@ -10,3 +10,7 @@
...
 )
+
+CONTAINERSTORAGE_API_V1 = EdgeResourceApi(
+    group="containerstorage.azure.com", version="v1", moniker="containerstorage"
```
[the remark this answers is not in the document]
Comment (Elsie4ever) [reply azext_edge/edge/providers/edge_api/arccontainerstorage.py:15]: I thought the moniker should follow the api group prefix?

File: azext_edge/edge/providers/support/base.py (hunk 1 above)
Comment (Elsie4ever) [reply azext_edge/edge/providers/support/base.py:80]: yes, will update!

File: azext_edge/edge/providers/edge_api/arccontainerstorage.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (digimaun) [reply azext_edge/edge/providers/edge_api/arccontainerstorage.py:15]: The moniker is supposed to be a human friendly alias that identifies the context of what is being captured.

File: azext_edge/tests/edge/support/create_bundle_int/test_arccontainerstorage_int.py (hunk 3 above)
Comment (Elsie4ever) [reply azext_edge/tests/edge/support/create_bundle_int/test_arccontainerstorage_int.py:96]: do we not want to test for more specific prefixes?

File: azext_edge/edge/providers/edge_api/arccontainerstorage.py (hunk 2 above)
Comment (Elsie4ever) [reply azext_edge/edge/providers/edge_api/arccontainerstorage.py:11]: yeah they both exist in the cluster

File: azext_edge/edge/_help.py [hunk 5]
```diff
@@ -56,7 +56,8 @@ def load_iotops_help():
             - {COMPAT_DEVICEREGISTRY_APIS.as_str()}
             - {COMPAT_CLUSTER_CONFIG_APIS.as_str()}
             - {COMPAT_DATAFLOW_APIS.as_str()}
-            - {COMPAT_ARCCONTAINERSTORAGE_APIS.as_str()}
+            - {ARCCONTAINERSTORAGE_API_V1.as_str()}
+            - {CONTAINERSTORAGE_API_V1.as_str()}
             - {SECRETSYNC_API_V1.as_str()}
```
[the remark this answers is not in the document]
Comment (Elsie4ever) [reply azext_edge/edge/_help.py:61]: when there is more than one api under COMPAT_<service>_APIS, I see the as_str() function adds a "\n" for each entry, which will cause runtime errors such as:
The command failed with an unexpected error. Here is the traceback:
expected '<document start>', but found '<scalar>'
  in "<unicode string>", line 14, column 1:
    containerstorage.azure.com/[v1]
    ^

[the remark this answers is not in the document]
Comment (digimaun) [reply azext_edge/edge/_help.py:61]: Ok sounds like a bug that needs to be fixed.

File: azext_edge/edge/providers/edge_api/arccontainerstorage.py (hunk 2 above)
Comment (Elsie4ever) [reply azext_edge/edge/providers/edge_api/arccontainerstorage.py:11]: we actually have another example with both api group exist, which is secret store:

secret-sync.x-k8s.io/v1alpha1
secrets-store.csi.x-k8s.io/v1

Comment (digimaun) [reply azext_edge/edge/providers/edge_api/arccontainerstorage.py:11]: Yea but those are both available on deployment with default settings.

Comment (digimaun) [reply azext_edge/edge/providers/edge_api/arccontainerstorage.py:11]: ^-- I'm basically asking if containerstorage.azure.com is applicable to the azure arc version of container storage, as we are focused on azure arc services.

File: azext_edge/tests/edge/support/create_bundle_int/helpers.py [hunk 6]
```diff
@@ -268,7 +269,13 @@ def get_file_map(
         file_map["usage"] = convert_file_names(walk_result[c_path]["files"])
         file_map["__namespaces__"]["usage"] = c_namespace
     elif ops_service == "acs":
-        assert len(walk_result) == 1 + expected_default_walk_result
+        if not acstor_namespace:
```
Comment (vilit1) [inline azext_edge/tests/edge/support/create_bundle_int/helpers.py:272]: switch the if branches

also comment for the magic numbers (why are you adding 2 vs 1 vs nothing)

File: azext_edge/tests/edge/support/create_bundle_int/test_arccontainerstorage_int.py (hunk 3 above)
Comment (vilit1) [reply azext_edge/tests/edge/support/create_bundle_int/test_arccontainerstorage_int.py:96]: this size of prefixes to check is getting a bit big - so if possible, why simplify/shorten the list

File: azext_edge/edge/_help.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Elsie4ever) [reply azext_edge/edge/_help.py:61]: will address this later https://msazure.visualstudio.com/One/_sprints/taskboard/IoT-UPX/One/Selenium/CY24Q4/2Wk/2Wk06%20(Dec%2015%20-%20Dec%2028)?workitem=30542801

File: azext_edge/edge/providers/edge_api/arccontainerstorage.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Elsie4ever) [reply azext_edge/edge/providers/edge_api/arccontainerstorage.py:15]: discussed offline, will keep as it is now

File: azext_edge/tests/edge/support/create_bundle_int/test_arccontainerstorage_int.py (hunk 3 above)
Comment (Elsie4ever) [reply azext_edge/tests/edge/support/create_bundle_int/test_arccontainerstorage_int.py]: updated

**Review by vilit1 — APPROVED** [review]
lgtm

**Pull request merged**